dotnet ecosystem: fix cross compilation

IvarWithoutBones opened the pull request:
###### Description of changes
Attempting to cross compile dotnet packages from for example `pkgsCross.aarch64-multiplatform.depotdownloader` used to not work for a few reasons, first it would fail because `makeWrapper` was in `makeSetupHook`'s `deps` attribute, which maps to `propagatedBuildInputs` instead of `nativeBuildInputs`.

Then, for whatever reason dotnet would fail to load `libstdc++.so.6` even though it was manually `patchelf`'d into its rpath, patching it in with `autoPatchelfHook` fixed that. Note that we cannot supply all dependencies through `autoPatchelfHook` since a few of them are `dlopen`'d which didn't seem to play nicely with it.

~~It also disables dotnet's globalisation support, which may help with deterministic builds. I would've expected this to be disabled already with the `Deterministic` msbuild property, but I noticed cross compiled binaries failed to run with globalisation support enabled.~~ Not required anymore.

cc @SuperSandro2000 @jonringer 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

IvarWithoutBones committed:
dotnet: fix cross compilation

IvarWithoutBones committed:
buildDotnetModule: fix cross compilation

IvarWithoutBones commented:
Fixed a regression with `github-runner` failing some unit tests by adding `DOTNET_SYSTEM_GLOBALIZATION_PREDEFINED_CULTURES_ONLY` to dotnet's setup hook :+1:

Mindavi commented on pkgs/build-support/dotnet/build-dotnet-module/hooks/default.nix:
```diff
@@ -53,7 +53,7 @@
   dotnetFixupHook = callPackage ({ }:
     makeSetupHook {
       name = "dotnet-fixup-hook";
-      deps = [ dotnet-runtime makeWrapper ];
```
Is makeWrapper unused in .net packages?

IvarWithoutBones replied:
It isn't, but `makeSetupHook`'s `deps` attribute maps to `propagatedBuildInputs`: https://github.com/NixOS/nixpkgs/blob/2c7a74c9925187476c7cd4f5dd227742b0c9a19a/pkgs/build-support/trivial-builders.nix#L537

That causes the following error:
```console
~/nix/nixpkgs > nix build -Lf . pkgsCross.aarch64-multiplatform.depotdownloader
error: attribute 'runtimeShell' missing

       at /home/ivv/nix/nixpkgs/pkgs/top-level/all-packages.nix:784:17:

          783|       substitutions = {
          784|         shell = targetPackages.runtimeShell;
             |                 ^
          785|         passthru.tests = tests.makeWrapper;
```
`makeWrapper` is also already provided through `buildDotnetModule`'s `nativeBuildInputs`: https://github.com/NixOS/nixpkgs/blob/2c7a74c9925187476c7cd4f5dd227742b0c9a19a/pkgs/build-support/dotnet/build-dotnet-module/default.nix#L99

Mindavi commented on pkgs/build-support/dotnet/build-dotnet-module/hooks/default.nix:
```diff
@@ -53,7 +53,7 @@
   dotnetFixupHook = callPackage ({ }:
     makeSetupHook {
       name = "dotnet-fixup-hook";
-      deps = [ dotnet-runtime makeWrapper ];
```
Is makeWrapper unused in .net packages?

Mindavi commented:
Seems ok other than the makeWrapper removal which needs a bit of explanation.

The regression of jackett is new? Maybe override the variables to the old ones for that package?

Mindavi commented:
Seems ok other than the makeWrapper removal which needs a bit of explanation.

The regression of jackett is new? Maybe override the variables to the old ones for that package?

SuperSandro2000 commented on pkgs/development/compilers/dotnet/build-dotnet.nix:
```diff
@@ -47,6 +48,10 @@ in stdenv.mkDerivation rec {
     lttng-ust_2_12
   ]);
 
+  # Fixes cross
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/compilers/dotnet/build-dotnet.nix:
```diff
@@ -47,6 +48,10 @@ in stdenv.mkDerivation rec {
   ]);
 
+  # Fixes cross
+  nativeBuildInputs = [ autoPatchelfHook ];
```
Why do we need to autopatchelf when compiling from source?

IvarWithoutBones replied:
I'm not entirely sure why, but without it a cross compiled binary fails as follows:
```console
~/nix/nixpkgs > nix build -Lf . pkgsCross.aarch64-multiplatform.dotnet-sdk_6
~/nix/nixpkgs > ./result/bin/dotnet
./result/bin/dotnet: error while loading shared libraries: libstdc++.so.6: cannot open shared object file: No such file or directory
```

SuperSandro2000 replied:
hmmm. I am not sure we should run patchelf on .NET binaries. That usually causes trouble.

IvarWithoutBones replied:
I haven't experienced any issues with it so far, do you know of any specific problems?

IvarWithoutBones commented:
> The regression of jackett is new? Maybe override the variables to the old ones for that package?

The issue is the msbuild flag, which does seem to be required for other packages. I've added a flag to disable the flag, and added it accordingly for jackett :+1:

mdarocha commented on pkgs/build-support/dotnet/build-dotnet-module/hooks/dotnet-configure-hook.sh:
```diff
@@ -13,6 +13,11 @@ dotnetConfigureHook() {
 
+    # Required for cross compiling, but breaks at least one package
+    if ! [ "${dontDotnetGlobalizationInvariantFlag}" ]; then
+        dotnetFlags+=("-p:InvariantGlobalization=true")
```
Since its used only on one package, isnt it better to set dotnetFlags there instead of adding another config option to buildDotnetModule?

IvarWithoutBones replied:
The flag is used for all packages except the ones enabling this flag though, not sure what a better solution would be.

mdarocha replied:
Oh, right, I misunderstood that part.

Still, disabling globalization has some implications for application behaviour (https://github.com/dotnet/runtime/blob/main/docs/design/features/globalization-invariant-mode.md). I'm not sure enabling it globally is a good idea, since it may introduce application bugs specific to NixOS, that may be undetected if the application tests don't cover invariant mode cases.

I'm not sure what a better alternative is. Is there a way to add some sort of warning during the build or a documentation note? That way at least someone packaging a new application will have some information if they run into globalization-specific bugs.

IvarWithoutBones replied:
> Still, disabling globalization has some implications for application behaviour (https://github.com/dotnet/runtime/blob/main/docs/design/features/globalization-invariant-mode.md). I'm not sure enabling it globally is a good idea, since it may introduce application bugs specific to NixOS, that may be undetected if the application tests don't cover invariant mode cases.

That's a good point, I hadn't considered that before. After looking into this further I found that we can just use the `icu` package used in `build-dotnet.nix` and wrap it into `LD_LIBRARY_PATH`, that works on both the natively and cross compiled binaries. With that we don't need this flag anymore :+1:

Mindavi reviewed (approved):
Running a nixpkgs-review but this LGTM.

Mindavi commented:
nixpkgs-review output on x86_64-linux:

error: builder for '/nix/store/nz99j0m20azy1xk3h2p391g1gzhkd7gg-netcoredbg-2.0.0-895.drv' failed with exit code 2;
       last 10 log lines:
       >    at System.Globalization.CultureInfo.get_CurrentUICulture()
       >    at System.TimeZoneInfo.GetUtcStandardDisplayName()
       >    at System.TimeZoneInfo.CreateUtcTimeZone()
       >    at System.TimeZoneInfo..cctor()
       >    at System.DateTime.get_Now()
       >    at Microsoft.DotNet.Cli.Program.Main(System.String[])
       > Subprocess aborted
       > make[2]: *** [src/CMakeFiles/netcoredbg.dir/build.make:86: src/generrmsg/generrmsg.csproj] Error 1
       > make[1]: *** [CMakeFiles/Makefile2:226: src/CMakeFiles/netcoredbg.dir/all] Error 2
       > make: *** [Makefile:156: all] Error 2
       For full logs, run 'nix log /nix/store/nz99j0m20azy1xk3h2p391g1gzhkd7gg-netcoredbg-2.0.0-895.drv'.
error: 1 dependencies of derivation '/nix/store/32l77a57p9xn1i24823h45ljafxhyzz2-netcoredbg-2.0.0-895.drv' failed to build
error: 1 dependencies of derivation '/nix/store/z1ddfryxv3iq0gaqr7vz3q49rgjibn67-review-shell.drv' failed to build

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/173889

1 package blacklisted:
tests.writers

1 package failed to build:
netcoredbg

39 packages built:
ArchiSteamFarm alttpr-opentracker azure-functions-core-tools baget btcpayserver depotdownloader discordchatexporter-cli dotnet-aspnetcore dotnet-runtime dotnet-sdk dotnet-sdk_3 dotnet-sdk_5 dotnetCorePackages.aspnetcore_3_1 dotnetCorePackages.aspnetcore_5_0 dotnetCorePackages.runtime_3_1 dotnetCorePackages.runtime_5_0 eventstore formula github-runner inklecate jackett jellyfin jetbrains.rider msbuild nbxplorer omnisharp-roslyn opentabletdriver osu-lazer pinta pre-commit prowlarr python-language-server radarr roslyn ryujinx vscode-extensions.ms-vsliveshare.vsliveshare wasabibackend wasabiwallet xivlauncher


After that it fails due to unfree packages.

First build fail probably same as: https://github.com/NixOS/nixpkgs/pull/181497#pullrequestreview-1039538365

lovesegfault commented:
This broke `pre-commit` on aarch64-darwin by means of breaking `dotnet-sdk` for all of darwin due to bringing in `autopatchelf`, which is Linux only.

lovesegfault commented:
I have a fix, it will be up in a couple of minutes

Mindavi commented:
Uh oh, sorry about that
